### PR TITLE
util, executor: use UnionRanges build index kv range for INLJ

### LINKED ISSUE #15686: executor: indexXXXJoin error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. What did you do?
download the file with all DDLs 
[randgen_index_merge_join.txt](https://github.com/pingcap/tidb/files/4381467/randgen_index_merge_join.txt)

```
source randgen_index_merge_join.txt
```                                             
```
SELECT /*+INL_MERGE_JOIN(table1,table2,table3,table4,table5,table6,table7,table8,table9,table10,table11,table12)*/ table1 . `col_int` AS field1 , table1 . `pk` AS field2 , table2 . `col_int` AS field3 , table2 . `col_int` AS field4 , table2 . `col_int` AS field5 , table1 . `col_int` AS field6 FROM  K AS table1  LEFT  JOIN A AS table2 ON  table1 . `col_int_key` =  table2 . `col_int_key`  WHERE table2 . `pk` < table1 . `pk`   ORDER BY field1, field2,field3,field4,field5,field6 ;
```

### 2. What did you expect to see?
**mysql is right, and using hash join also is right, but indexjoin and indexHashJoin are also wrong.**
```
+--------+--------+--------+--------+--------+--------+
| field1 | field2 | field3 | field4 | field5 | field6 |
+--------+--------+--------+--------+--------+--------+
|   NULL |     14 |      1 |      1 |      1 |   NULL |
|   NULL |     28 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     28 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     28 |      6 |      6 |      6 |   NULL |
|   NULL |     28 |      7 |      7 |      7 |   NULL |
|   NULL |     28 |      8 |      8 |      8 |   NULL |
|   NULL |     33 |      7 |      7 |      7 |   NULL |
|   NULL |     35 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     35 |      6 |      6 |      6 |   NULL |
|   NULL |     46 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     46 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     46 |      3 |      3 |      3 |   NULL |
|   NULL |     46 |      6 |      6 |      6 |   NULL |
|   NULL |     46 |      7 |      7 |      7 |   NULL |
|   NULL |     46 |      8 |      8 |      8 |   NULL |
|   NULL |     60 |      0 |      0 |      0 |   NULL |
|   NULL |     60 |      2 |      2 |      2 |   NULL |
|   NULL |     60 |      3 |      3 |      3 |   NULL |
|   NULL |     60 |      3 |      3 |      3 |   NULL |
|   NULL |     60 |      5 |      5 |      5 |   NULL |
|   NULL |     60 |      6 |      6 |      6 |   NULL |
|   NULL |     66 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     66 |      4 |      4 |      4 |   NULL |
|      0 |     36 |      0 |      0 |      0 |      0 |
|      0 |     36 |      2 |      2 |      2 |      0 |
|      0 |     45 |   NULL |   NULL |   NULL |      0 |
|      0 |     45 |      1 |      1 |      1 |      0 |
|      0 |     45 |      2 |      2 |      2 |      0 |
|      0 |     45 |      3 |      3 |      3 |      0 |
|      0 |     45 |      6 |      6 |      6 |      0 |
|      1 |     23 |      6 |      6 |      6 |      1 |
|      1 |     34 |      4 |      4 |      4 |      1 |
|      1 |     52 |   NULL |   NULL |   NULL |      1 |
|      1 |     52 |      2 |      2 |      2 |      1 |
|      1 |     52 |      5 |      5 |      5 |      1 |
|      1 |     52 |      7 |      7 |      7 |      1 |
|      1 |     57 |      0 |      0 |      0 |      1 |
|      1 |     57 |      2 |      2 |      2 |      1 |
|      1 |     57 |      3 |      3 |      3 |      1 |
|      1 |     57 |      3 |      3 |      3 |      1 |
|      1 |     57 |      5 |      5 |      5 |      1 |
|      1 |     57 |      6 |      6 |      6 |      1 |
|      1 |     64 |      1 |      1 |      1 |      1 |
|      1 |     64 |      1 |      1 |      1 |      1 |
|      1 |     64 |      2 |      2 |      2 |      1 |
|      1 |     69 |   NULL |   NULL |   NULL |      1 |
|      1 |     69 |   NULL |   NULL |   NULL |      1 |
|      1 |     69 |      2 |      2 |      2 |      1 |
|      1 |     69 |      5 |      5 |      5 |      1 |
|      1 |     69 |      7 |      7 |      7 |      1 |
|      1 |     69 |      7 |      7 |      7 |      1 |
|      2 |     17 |      4 |      4 |      4 |      2 |
|      2 |     25 |      0 |      0 |      0 |      2 |
|      2 |     25 |      3 |      3 |      3 |      2 |
|      2 |     40 |   NULL |   NULL |   NULL |      2 |
|      2 |     40 |      4 |      4 |      4 |      2 |
|      2 |     42 |      0 |      0 |      0 |      2 |
|      2 |     42 |      2 |      2 |      2 |      2 |
|      2 |     42 |      3 |      3 |      3 |      2 |
|      2 |     61 |      0 |      0 |      0 |      2 |
|      2 |     61 |      8 |      8 |      8 |      2 |
|      3 |     65 |   NULL |   NULL |   NULL |      3 |
|      3 |     65 |   NULL |   NULL |   NULL |      3 |
|      3 |     65 |      3 |      3 |      3 |      3 |
|      3 |     65 |      5 |      5 |      5 |      3 |
|      3 |     65 |      6 |      6 |      6 |      3 |
|      3 |     65 |      7 |      7 |      7 |      3 |
|      3 |     65 |      8 |      8 |      8 |      3 |
|      3 |     68 |      0 |      0 |      0 |      3 |
|      3 |     68 |      2 |      2 |      2 |      3 |
|      3 |     68 |      3 |      3 |      3 |      3 |
|      3 |     68 |      3 |      3 |      3 |      3 |
|      3 |     68 |      5 |      5 |      5 |      3 |
|      3 |     68 |      6 |      6 |      6 |      3 |
|      4 |     32 |   NULL |   NULL |   NULL |      4 |
|      4 |     32 |   NULL |   NULL |   NULL |      4 |
|      4 |     32 |      6 |      6 |      6 |      4 |
|      4 |     32 |      7 |      7 |      7 |      4 |
|      4 |     32 |      8 |      8 |      8 |      4 |
|      5 |     43 |      1 |      1 |      1 |      5 |
|      5 |     43 |      1 |      1 |      1 |      5 |
|      5 |     43 |      2 |      2 |      2 |      5 |
|      5 |     48 |      0 |      0 |      0 |      5 |
|      5 |     48 |      1 |      1 |      1 |      5 |
|      5 |     48 |      3 |      3 |      3 |      5 |
|      5 |     48 |      4 |      4 |      4 |      5 |
|      5 |     59 |      8 |      8 |      8 |      5 |
|      6 |     26 |      6 |      6 |      6 |      6 |
|      6 |     67 |   NULL |   NULL |   NULL |      6 |
|      6 |     67 |      4 |      4 |      4 |      6 |
|      7 |     47 |   NULL |   NULL |   NULL |      7 |
|      7 |     47 |      1 |      1 |      1 |      7 |
|      7 |     47 |      2 |      2 |      2 |      7 |
|      7 |     47 |      3 |      3 |      3 |      7 |
|      7 |     47 |      6 |      6 |      6 |      7 |
|      7 |     49 |   NULL |   NULL |   NULL |      7 |
|      7 |     49 |      1 |      1 |      1 |      7 |
|      7 |     49 |      2 |      2 |      2 |      7 |
|      7 |     49 |      2 |      2 |      2 |      7 |
|      7 |     49 |      3 |      3 |      3 |      7 |
|      7 |     49 |      6 |      6 |      6 |      7 |
|      7 |     70 |      0 |      0 |      0 |      7 |
|      7 |     70 |      7 |      7 |      7 |      7 |
|      7 |     70 |      8 |      8 |      8 |      7 |
|      8 |      9 |      0 |      0 |      0 |      8 |
|      8 |      9 |      3 |      3 |      3 |      8 |
|      8 |     16 |      1 |      1 |      1 |      8 |
|      8 |     31 |      6 |      6 |      6 |      8 |
|      8 |     39 |   NULL |   NULL |   NULL |      8 |
|      8 |     39 |      4 |      4 |      4 |      8 |
|      8 |     41 |      0 |      0 |      0 |      8 |
|      8 |     41 |      1 |      1 |      1 |      8 |
|      8 |     41 |      3 |      3 |      3 |      8 |
|      8 |     41 |      4 |      4 |      4 |      8 |
|      9 |     12 |      1 |      1 |      1 |      9 |
|      9 |     13 |   NULL |   NULL |   NULL |      9 |
|      9 |     13 |      6 |      6 |      6 |      9 |
|      9 |     13 |      7 |      7 |      7 |      9 |
|      9 |     22 |   NULL |   NULL |   NULL |      9 |
|      9 |     22 |   NULL |   NULL |   NULL |      9 |
|      9 |     22 |      6 |      6 |      6 |      9 |
|      9 |     22 |      7 |      7 |      7 |      9 |
|      9 |     22 |      8 |      8 |      8 |      9 |
|      9 |     27 |      6 |      6 |      6 |      9 |
|      9 |     56 |   NULL |   NULL |   NULL |      9 |
|      9 |     56 |   NULL |   NULL |   NULL |      9 |
|      9 |     56 |      3 |      3 |      3 |      9 |
|      9 |     56 |      6 |      6 |      6 |      9 |
|      9 |     56 |      7 |      7 |      7 |      9 |
|      9 |     56 |      8 |      8 |      8 |      9 |
|      9 |     58 |      1 |      1 |      1 |      9 |
|      9 |     58 |      1 |      1 |      1 |      9 |
|      9 |     58 |      2 |      2 |      2 |      9 |
|      9 |     62 |   NULL |   NULL |   NULL |      9 |
|      9 |     62 |   NULL |   NULL |   NULL |      9 |
|      9 |     62 |      2 |      2 |      2 |      9 |
|      9 |     62 |      5 |      5 |      5 |      9 |
|      9 |     62 |      7 |      7 |      7 |      9 |
|      9 |     62 |      7 |      7 |      7 |      9 |
|      9 |     63 |   NULL |   NULL |   NULL |      9 |
|      9 |     63 |   NULL |   NULL |   NULL |      9 |
|      9 |     63 |      2 |      2 |      2 |      9 |
|      9 |     63 |      5 |      5 |      5 |      9 |
|      9 |     63 |      7 |      7 |      7 |      9 |
|      9 |     63 |      7 |      7 |      7 |      9 |
+--------+--------+--------+--------+--------+--------+
145 rows in set, 1 warning (0.01 sec)

```
### 3. What did you see instead?
```
+--------+--------+--------+--------+--------+--------+
| field1 | field2 | field3 | field4 | field5 | field6 |
+--------+--------+--------+--------+--------+--------+
|   NULL |     28 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     28 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     28 |      6 |      6 |      6 |   NULL |
|   NULL |     28 |      7 |      7 |      7 |   NULL |
|   NULL |     28 |      8 |      8 |      8 |   NULL |
|   NULL |     33 |      7 |      7 |      7 |   NULL |
|   NULL |     35 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     35 |      6 |      6 |      6 |   NULL |
|   NULL |     46 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     46 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     46 |      3 |      3 |      3 |   NULL |
|   NULL |     46 |      6 |      6 |      6 |   NULL |
|   NULL |     46 |      7 |      7 |      7 |   NULL |
|   NULL |     46 |      8 |      8 |      8 |   NULL |
|   NULL |     60 |      0 |      0 |      0 |   NULL |
|   NULL |     60 |      2 |      2 |      2 |   NULL |
|   NULL |     60 |      3 |      3 |      3 |   NULL |
|   NULL |     60 |      3 |      3 |      3 |   NULL |
|   NULL |     60 |      5 |      5 |      5 |   NULL |
|   NULL |     60 |      6 |      6 |      6 |   NULL |
|   NULL |     66 |   NULL |   NULL |   NULL |   NULL |
|   NULL |     66 |      4 |      4 |      4 |   NULL |
|      0 |     36 |      0 |      0 |      0 |      0 |
|      0 |     36 |      2 |      2 |      2 |      0 |
|      0 |     45 |   NULL |   NULL |   NULL |      0 |
|      0 |     45 |      1 |      1 |      1 |      0 |
|      0 |     45 |      2 |      2 |      2 |      0 |
|      0 |     45 |      3 |      3 |      3 |      0 |
|      0 |     45 |      6 |      6 |      6 |      0 |
|      1 |     23 |      6 |      6 |      6 |      1 |
|      1 |     34 |      4 |      4 |      4 |      1 |
|      1 |     52 |   NULL |   NULL |   NULL |      1 |
|      1 |     52 |      2 |      2 |      2 |      1 |
|      1 |     52 |      5 |      5 |      5 |      1 |
|      1 |     52 |      7 |      7 |      7 |      1 |
|      1 |     57 |      0 |      0 |      0 |      1 |
|      1 |     57 |      2 |      2 |      2 |      1 |
|      1 |     57 |      3 |      3 |      3 |      1 |
|      1 |     57 |      3 |      3 |      3 |      1 |
|      1 |     57 |      5 |      5 |      5 |      1 |
|      1 |     57 |      6 |      6 |      6 |      1 |
|      1 |     69 |   NULL |   NULL |   NULL |      1 |
|      1 |     69 |   NULL |   NULL |   NULL |      1 |
|      1 |     69 |      2 |      2 |      2 |      1 |
|      1 |     69 |      5 |      5 |      5 |      1 |
|      1 |     69 |      7 |      7 |      7 |      1 |
|      1 |     69 |      7 |      7 |      7 |      1 |
|      2 |     17 |      4 |      4 |      4 |      2 |
|      2 |     25 |      0 |      0 |      0 |      2 |
|      2 |     25 |      3 |      3 |      3 |      2 |
|      2 |     40 |   NULL |   NULL |   NULL |      2 |
|      2 |     40 |      4 |      4 |      4 |      2 |
|      2 |     42 |      0 |      0 |      0 |      2 |
|      2 |     42 |      2 |      2 |      2 |      2 |
|      2 |     42 |      3 |      3 |      3 |      2 |
|      2 |     61 |      0 |      0 |      0 |      2 |
|      2 |     61 |      8 |      8 |      8 |      2 |
|      3 |     65 |   NULL |   NULL |   NULL |      3 |
|      3 |     65 |   NULL |   NULL |   NULL |      3 |
|      3 |     65 |      3 |      3 |      3 |      3 |
|      3 |     65 |      5 |      5 |      5 |      3 |
|      3 |     65 |      6 |      6 |      6 |      3 |
|      3 |     65 |      7 |      7 |      7 |      3 |
|      3 |     65 |      8 |      8 |      8 |      3 |
|      3 |     68 |      0 |      0 |      0 |      3 |
|      3 |     68 |      2 |      2 |      2 |      3 |
|      3 |     68 |      3 |      3 |      3 |      3 |
|      3 |     68 |      3 |      3 |      3 |      3 |
|      3 |     68 |      5 |      5 |      5 |      3 |
|      3 |     68 |      6 |      6 |      6 |      3 |
|      4 |     32 |   NULL |   NULL |   NULL |      4 |
|      4 |     32 |   NULL |   NULL |   NULL |      4 |
|      4 |     32 |      6 |      6 |      6 |      4 |
|      4 |     32 |      7 |      7 |      7 |      4 |
|      4 |     32 |      8 |      8 |      8 |      4 |
|      5 |     48 |      0 |      0 |      0 |      5 |
|      5 |     48 |      1 |      1 |      1 |      5 |
|      5 |     48 |      3 |      3 |      3 |      5 |
|      5 |     48 |      4 |      4 |      4 |      5 |
|      5 |     59 |      8 |      8 |      8 |      5 |
|      6 |     26 |      6 |      6 |      6 |      6 |
|      6 |     67 |   NULL |   NULL |   NULL |      6 |
|      6 |     67 |      4 |      4 |      4 |      6 |
|      7 |     47 |   NULL |   NULL |   NULL |      7 |
|      7 |     47 |      1 |      1 |      1 |      7 |
|      7 |     47 |      2 |      2 |      2 |      7 |
|      7 |     47 |      3 |      3 |      3 |      7 |
|      7 |     47 |      6 |      6 |      6 |      7 |
|      7 |     49 |   NULL |   NULL |   NULL |      7 |
|      7 |     49 |      1 |      1 |      1 |      7 |
|      7 |     49 |      2 |      2 |      2 |      7 |
|      7 |     49 |      2 |      2 |      2 |      7 |
|      7 |     49 |      3 |      3 |      3 |      7 |
|      7 |     49 |      6 |      6 |      6 |      7 |
|      7 |     70 |      0 |      0 |      0 |      7 |
|      7 |     70 |      7 |      7 |      7 |      7 |
|      7 |     70 |      8 |      8 |      8 |      7 |
|      8 |      9 |      0 |      0 |      0 |      8 |
|      8 |      9 |      3 |      3 |      3 |      8 |
|      8 |     31 |      6 |      6 |      6 |      8 |
|      8 |     39 |   NULL |   NULL |   NULL |      8 |
|      8 |     39 |      4 |      4 |      4 |      8 |
|      8 |     41 |      0 |      0 |      0 |      8 |
|      8 |     41 |      1 |      1 |      1 |      8 |
|      8 |     41 |      3 |      3 |      3 |      8 |
|      8 |     41 |      4 |      4 |      4 |      8 |
|      9 |     13 |   NULL |   NULL |   NULL |      9 |
|      9 |     13 |      6 |      6 |      6 |      9 |
|      9 |     13 |      7 |      7 |      7 |      9 |
|      9 |     22 |   NULL |   NULL |   NULL |      9 |
|      9 |     22 |   NULL |   NULL |   NULL |      9 |
|      9 |     22 |      6 |      6 |      6 |      9 |
|      9 |     22 |      7 |      7 |      7 |      9 |
|      9 |     22 |      8 |      8 |      8 |      9 |
|      9 |     27 |      6 |      6 |      6 |      9 |
|      9 |     56 |   NULL |   NULL |   NULL |      9 |
|      9 |     56 |   NULL |   NULL |   NULL |      9 |
|      9 |     56 |      3 |      3 |      3 |      9 |
|      9 |     56 |      6 |      6 |      6 |      9 |
|      9 |     56 |      7 |      7 |      7 |      9 |
|      9 |     56 |      8 |      8 |      8 |      9 |
|      9 |     62 |   NULL |   NULL |   NULL |      9 |
|      9 |     62 |   NULL |   NULL |   NULL |      9 |
|      9 |     62 |      2 |      2 |      2 |      9 |
|      9 |     62 |      5 |      5 |      5 |      9 |
|      9 |     62 |      7 |      7 |      7 |      9 |
|      9 |     62 |      7 |      7 |      7 |      9 |
|      9 |     63 |   NULL |   NULL |   NULL |      9 |
|      9 |     63 |   NULL |   NULL |   NULL |      9 |
|      9 |     63 |      2 |      2 |      2 |      9 |
|      9 |     63 |      5 |      5 |      5 |      9 |
|      9 |     63 |      7 |      7 |      7 |      9 |
|      9 |     63 |      7 |      7 |      7 |      9 |
+--------+--------+--------+--------+--------+--------+
133 rows in set, 2 warnings (0.01 sec)

```
### 4. What version of TiDB are you using? (`tidb-server -V` or run `select tidb_version();` on TiDB)
| 5.7.25-TiDB-v4.0.0-beta-523-g7eba696bb-dirty |

### PR DESCRIPTION
**Pull request by XuHuaiyu**
### What problem does this PR solve?

Issue Number: close #15686  
close https://github.com/pingcap/tidb/issues/15693
close https://github.com/pingcap/tidb/issues/15687

Problem Summary:
Before this commit, we build a wrong kv ranges which will cause data lost when index join fetchInnerResult. 

### What is changed and how it works?

What's Changed:
Use UnionRanges to merge ranges.

How it Works:

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

- Performance regression
    - Consumes more CPU

### CODE REVIEW

**Comment by XuHuaiyu** [comment]
PTAL @winoros @eurekaka

File: util/ranger/ranger.go [hunk 1]
```diff
@@ -356,7 +356,11 @@ type sortRange struct {
 	encodedEnd    []byte
 }
 
-func unionRanges(sc *stmtctx.StatementContext, ranges []*Range) ([]*Range, error) {
+// UnionRanges sorts `ranges`, union adjacent ones if possible.
+// For two intervals [a, b], [c, d], we have guaranteed that a >= c. If b >= c. Then two intervals are overlapped.
```
Comment (eurekaka) [inline util/ranger/ranger.go]:
```suggestion
// For two intervals [a, b], [c, d], we have guaranteed that a <= c. If b >= c. Then two intervals are overlapped.
```
?

**Comment by winoros** [comment]
I cannot figure out why the old codes is wrong since the merge algo is the same. Can you create a issue for future opotimization?

**Comment by XuHuaiyu** [comment]
unit test failed because of https://github.com/pingcap/tidb/issues/15594

**Pull request merged**